<a href="https://colab.research.google.com/github/Suhail-46/Asteroid-Tracker-for-Earth/blob/main/NASA_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
pip install mysql-connector-python # installation of SQL connector

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 33.9/33.9 MB 38.7 MB/s eta 0:00:00


In [2]:
# Code used to scrape 10000 data from NASA API
from datetime import datetime
import requests
import pprint
api_key = "zVrrXmXMHFlLP3ow0eNmYfJ2RWaMrjyKJi9eBYvl"
asteroids_data = [] # Created empty list to store the collection of values from scraping
target = 10000
url = f"https://api.nasa.gov/neo/rest/v1/feed?start_date=2024-01-01&end_date=2024-01-07&api_key={api_key}" # Request given to collect data within date range of 2024-01-01 to 2024-01-07
response = requests.get(url)
data = response.json()
details = data["near_earth_objects"]
while len(asteroids_data) < target: # Loop used to collect 10000 data
  for date, ast_details in details.items(): # Loop used to collect the data based on range of date
      for ast in ast_details: # Loop used to extract the field
        asteroids_data.append(dict(
            id = ast["id"],
            neo_reference_id = ast["neo_reference_id"],
            name = ast["name"],
            absolute_magnitude_h = ast["absolute_magnitude_h"],
            estimated_diameter_min_km = ast["estimated_diameter"]["kilometers"]["estimated_diameter_min"],
            estimated_diameter_max_km = ast["estimated_diameter"]["kilometers"]["estimated_diameter_max"],
            is_potentially_hazardous_asteroid = ast["is_potentially_hazardous_asteroid"],
            close_approach_date = datetime.strptime(ast["close_approach_data"][0]["close_approach_date"], "%Y-%m-%d"),
            relative_velocity_kmph = ast["close_approach_data"][0]["relative_velocity"]["kilometers_per_hour"],
            astronomical = ast["close_approach_data"][0]["miss_distance"]["astronomical"],
            miss_distance_km = ast["close_approach_data"][0]["miss_distance"]["kilometers"],
            miss_distance_lunar = ast["close_approach_data"][0]["miss_distance"]["lunar"],
            orbiting_body = ast["close_approach_data"][0]["orbiting_body"]
        ))
        if len(asteroids_data) == target:
            break # Loop breaks once all fields of data collected
      if len(asteroids_data) == target:
        break # Loop breaks once all data collected within the range of date
url = data["links"]["next"] # The loops were applied to the next page of API
# Two list of data is extracted to feed on the table as per the requirement
asteroids_list = [] # Created an empty list to store data of asteroids with respect to the requirement of asteroids table
for asteroids in asteroids_data:
    asteroids_list.append((
        asteroids['id'],
        asteroids['name'],
        asteroids['absolute_magnitude_h'],
        asteroids['estimated_diameter_min_km'],
        asteroids['estimated_diameter_max_km'],
        asteroids['is_potentially_hazardous_asteroid']
    ))
close_approach_list = [] # Created an empty list to store data of close approach with respect to the requirement of close_approach table
for close_approach in asteroids_data:
    close_approach_list.append((
        close_approach['neo_reference_id'],
        close_approach['close_approach_date'],
        close_approach['relative_velocity_kmph'],
        close_approach['astronomical'],
        close_approach['miss_distance_km'],
        close_approach['miss_distance_lunar'],
        close_approach['orbiting_body']
    ))

In [3]:
asteroids_list

[('2415949', '415949 (2001 XY10)', 19.37, 0.3552670883, 0.7944013596, False),
 ('3160747', '(2003 SR84)', 26.0, 0.0167708462, 0.0375007522, False),
 ('3309828', '(2005 YQ96)', 20.62, 0.1997813652, 0.4467247133, True),
 ('3457842', '(2009 HC21)', 22.1, 0.1010543415, 0.2259643771, False),
 ('3553062', '(2010 XA11)', 26.1, 0.0160160338, 0.0358129403, False),
 ('3591616', '(2011 YP10)', 23.94, 0.0433066885, 0.0968366995, False),
 ('3608936', '(2012 SD22)', 20.05, 0.2597496533, 0.580817882, True),
 ('3795154', '(2017 YD8)', 21.87, 0.1123453177, 0.2512117673, False),
 ('3842680', '(2019 KK5)', 22.79, 0.0735453089, 0.1644523102, False),
 ('2613286', '613286 (2005 YQ96)', 20.63, 0.1988634532, 0.4446721997, True),
 ('54338714', '(2023 AW)', 25.68, 0.0194336768, 0.0434550225, False),
 ('54415081', '(2023 XZ4)', 24.05, 0.0411675457, 0.0920534307, False),
 ('54417653', '(2023 YR)', 24.9, 0.0278326768, 0.0622357573, False),
 ('54418497', '(2024 AA)', 27.41, 0.0087610261, 0.01959025, False),
 ('5441

In [4]:
len(asteroids_list)

10000

In [5]:
len(close_approach_list)

10000

In [6]:
# Pushing the list of values into data frame for cleaning
import pandas as pd
df1 = pd.DataFrame(asteroids_list, columns=['id', 'name', 'absolute_magnitude_h', 'estimated_diameter_min_km', 'estimated_diameter_max_km', 'is_potentially_hazardous_asteroid'])
df2 = pd.DataFrame(close_approach_list, columns=['neo_reference_id', 'close_approach_date', 'relative_velocity_kmph', 'astronomical', 'miss_distance_km', 'miss_distance_lunar', 'orbiting_body'])

In [7]:
# Checking the data types are properly as int or float as per requirement
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   id                                 10000 non-null  object 
 1   name                               10000 non-null  object 
 2   absolute_magnitude_h               10000 non-null  float64
 3   estimated_diameter_min_km          10000 non-null  float64
 4   estimated_diameter_max_km          10000 non-null  float64
 5   is_potentially_hazardous_asteroid  10000 non-null  bool   
dtypes: bool(1), float64(3), object(2)
memory usage: 400.5+ KB


In [8]:
# Ensuring there are no null values
df1.isnull().sum()

id                                   0
name                                 0
absolute_magnitude_h                 0
estimated_diameter_min_km            0
estimated_diameter_max_km            0
is_potentially_hazardous_asteroid    0
dtype: int64

In [9]:
# Since the values of id column can be converted from string to int and name column contains characters which can't be converted as numerical data types so only id column is converted.
df1['id'] = df1['id'].astype('Int64')
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   id                                 10000 non-null  Int64  
 1   name                               10000 non-null  object 
 2   absolute_magnitude_h               10000 non-null  float64
 3   estimated_diameter_min_km          10000 non-null  float64
 4   estimated_diameter_max_km          10000 non-null  float64
 5   is_potentially_hazardous_asteroid  10000 non-null  bool   
dtypes: Int64(1), bool(1), float64(3), object(1)
memory usage: 410.3+ KB


In [10]:
# Checking duplicated rows
df1.duplicated().sum()

np.int64(9890)

In [11]:
# Checking the data types are properly as int or float as per requirement
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   neo_reference_id        10000 non-null  object        
 1   close_approach_date     10000 non-null  datetime64[ns]
 2   relative_velocity_kmph  10000 non-null  object        
 3   astronomical            10000 non-null  object        
 4   miss_distance_km        10000 non-null  object        
 5   miss_distance_lunar     10000 non-null  object        
 6   orbiting_body           10000 non-null  object        
dtypes: datetime64[ns](1), object(6)
memory usage: 547.0+ KB


In [12]:
# Displying the values in df2 datafreme(close_approach_list).
df2

neo_reference_id close_approach_date relative_velocity_kmph  \
0             2415949          2024-01-02       57205.8951204341   
1             3160747          2024-01-02        38589.054833182   
2             3309828          2024-01-02       56413.0143519451   
3             3457842          2024-01-02       21891.1182185894   
4             3553062          2024-01-02       31468.9783588978   
...               ...                 ...                    ...   
9995         54435948          2024-01-05        25849.764640261   
9996          2199003          2024-01-07       55047.0116924064   
9997          2434188          2024-01-07      121814.4318166937   
9998          3398095          2024-01-07       63326.0734302496   
9999          3520662          2024-01-07       38150.2228974764   

      astronomical    miss_distance_km miss_distance_lunar orbiting_body  
0     0.3372535274  50452409.349026638      131.1916221586         Earth  
1     0.1323425924  19798169.933318188       51.4812684436         Earth  
2     0.1670126223  24984732.559194501       64.9679100747         Earth  
3     0.4920511029  73609796.924990823      191.4078790281         Earth  
4     0.2358022486  35275514.131770482       91.7270747054         Earth  
...            ...                 ...                 ...           ...  
9995   0.337303243   50459846.69689241       131.210961527         Earth  
9996  0.1995345465  29849943.147815955       77.6189385885         Earth  
9997   0.204313989   30564937.56560343        79.478141721         Earth  
9998  0.1917667009  28687889.991567083       74.5972466501         Earth  
9999   0.203565311   30452936.93148757        79.186905979         Earth  

[10000 rows x 7 columns]

Since there are many sting data types and it can be converted while cretaing table and inserting in SQL. This is another method that we can use for convertion

In [13]:
# Ensuring there are no null values
df2.isnull().sum()

neo_reference_id          0
close_approach_date       0
relative_velocity_kmph    0
astronomical              0
miss_distance_km          0
miss_distance_lunar       0
orbiting_body             0
dtype: int64

In [14]:
# Ensuring there are no duplicated rows
df2.duplicated().sum()

np.int64(9890)

There are only 110(10000 - 9890) unique rows in the asteroid_list so we can't use the id column as primary key so while insertion data into SQL table ignoring the duplicted values by using INSERT IGNORE INTO method

In [15]:
# Code to create table and insert the bulk of data into the created table
import mysql.connector
suhail = mysql.connector.connect(host="gateway01.ap-southeast-1.prod.aws.tidbcloud.com",
                                 user="2XT5EKXhUVwb2JW.root",
                                 password="YO7YdnLJV098Hccw",
                                 port = 4000,
                                 database="NASA_Project") # Establish DB connection using TiDB cloud
connection = suhail.cursor()
connection.execute("create table asteroids (id int, name varchar(20), absolute_magnitude_h float, estimated_diameter_min_km float, estimated_diameter_max_km float, is_potentially_hazardous_asteroid boolean, constraint pk_id primary key (id))")
connection.execute("create table close_approach (neo_reference_id int, close_approach_date date, relative_velocity_kmph float, astronomical float, miss_distance_km float, miss_distance_lunar float, orbiting_body varchar(20), constraint fk_neo_reference_id foreign key (neo_reference_id) references asteroids(id))")
insert_query_1 = '''INSERT IGNORE INTO asteroids (id, name, absolute_magnitude_h, estimated_diameter_min_km, estimated_diameter_max_km, is_potentially_hazardous_asteroid) VALUES (%s, %s, %s, %s, %s, %s)'''
insert_query_2 = '''INSERT INTO close_approach (neo_reference_id, close_approach_date, relative_velocity_kmph, astronomical, miss_distance_km, miss_distance_lunar, orbiting_body) VALUES (%s, %s, %s, %s, %s, %s, %s)'''
connection.executemany(insert_query_1, asteroids_list)
connection.executemany(insert_query_2, close_approach_list)
suhail.commit()
print("Inserted asteroids and close_approach data into MySQL!")

Inserted asteroids and close_approach data into MySQL!


In [16]:
!pip install streamlit streamlit_option_menu # installation of streamlit for data visualization

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.3/44.3 kB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 61.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 829.3/829.3 kB 49.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 109.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.1/79.1 kB 7.3 MB/s eta 0:00:00


In [17]:
# Writing code in .py file format used to support streamlit
%%writefile app.py
import streamlit as st
import pandas as pd
import mysql.connector
import datetime
suhail = mysql.connector.connect(host="gateway01.ap-southeast-1.prod.aws.tidbcloud.com",
                                 user="2XT5EKXhUVwb2JW.root",
                                 password="YO7YdnLJV098Hccw",
                                 port = 4000,
                                 database="NASA_Project")
connection = suhail.cursor()
st.sidebar.title("🌠 Asteroid Approach")
st.sidebar.divider()
menu = st.sidebar.radio("Navigate",("Filter Criteria", "Queries"))
st.title("🚀 :blue[Asteroid Tracker for Earth] ☄️")
st.divider()
if menu == "Filter Criteria":
  left, middle, right = st.columns(3) # Organizing the data using sliders and select boxes
  min_mag = left.slider("Min Magnitude", 13.00, 32.61, (13.00, 32.61))
  velocity = middle.slider("Relative_velocity_kmph Range", 1418.21, 173071.83, (1418.21, 173071.83))
  min_dia = left.slider("Min Estimated Diameter(km)", 0.00, 4.62, (0.00, 4.62))
  max_dia = left.slider("Max Estimated Diameter(km)", 0.00, 10.33, (0.00, 10.33))
  astronomical = middle.slider("Astronomical unit", 0.00, 0.50, (0.00, 0.50))
  hazard = middle.selectbox("Only Show Potentially Hazardous", ("1", "0"))
  str_dt = right.date_input("Start Date", datetime.date(2024, 1, 1))
  end_dt = right.date_input("End Date", datetime.date(2024, 1, 7))
  button = st.button("Filter")
  if button:
    st.header("Filtered Asteroids", divider="blue")
    query = """select distinct a.id, a.name, a.absolute_magnitude_h, a.estimated_diameter_min_km, a.estimated_diameter_max_km,
              a.is_potentially_hazardous_asteroid, c.neo_reference_id, c.close_approach_date, c.relative_velocity_kmph, c.astronomical,
              c.miss_distance_km, c.miss_distance_lunar, c.orbiting_body
              from asteroids a
              join close_approach c on a.id = c.neo_reference_id
              where a.absolute_magnitude_h between %s and %s
              and a.estimated_diameter_min_km between %s and %s
              and a.estimated_diameter_max_km between %s and %s
              and c.relative_velocity_kmph between %s and %s
              and c.astronomical between %s and %s
              and c.close_approach_date between %s and %s
              and a.is_potentially_hazardous_asteroid = %s order by a.id""" # Query used to display data based on the applied filters over sliders and select box
    params = (min_mag[0], min_mag[1],
             min_dia[0], min_dia[1],
             max_dia[0], max_dia[1],
             velocity[0], velocity[1],
             astronomical[0], astronomical[1],
               str_dt, end_dt, int(hazard)) # Parameters are used to pass on based on the appiled filters
    connection.execute(query, params)
    records = connection.fetchall()
    df = pd.DataFrame(records, columns=[col[0] for col in connection.description])
    st.dataframe(df)
elif menu == "Queries": # Code written for another tab of side menu to diplay the quries based on the requirements
    option = st.selectbox(
        "Select your Query",
        (
            "Count of how many times each asteroid has approached Earth",
            "Average velocity of each asteroid over multiple approaches",
            "List top 10 fastest asteroids",
            "Potentially hazardous asteroids that have approached Earth more than 3 times",
            "Month with the most asteroid approaches",
            "Asteroid with the fastest ever approach speed",
            "Sorted Asteroids in descending order by maximum estimated diameter",
            "Asteroid whose closest approach is getting nearer over time",
            "Asteroids of its closest approach to Earth along with the date and miss distance",
            "Asteroids that approached Earth with velocity > 50,000 km/h",
            "Count of asteroids approaches happened per month",
            "Asteroid with the highest brightness",
            "Number of hazardous vs non-hazardous asteroids",
            "Asteroids that passed closer than the Moon (lesser than 1 LD), along with their close approach date and distance",
            "Asteroids that came within 0.05 AU"
        )
    )

    if option == "Count of how many times each asteroid has approached Earth":
        query1 = """select a.name, count(c.neo_reference_id) as count from asteroids a join close_approach c on a.id = c.neo_reference_id group by a.name"""
        df1 = pd.read_sql(query1, suhail)
        st.dataframe(df1)
    elif option == "Average velocity of each asteroid over multiple approaches":
        query2 = """select a.name, avg(c.relative_velocity_kmph) as average_velocity from asteroids a join close_approach c on a.id = c.neo_reference_id group by a.name"""
        df2 = pd.read_sql(query2, suhail)
        st.dataframe(df2)
    elif option == "List top 10 fastest asteroids":
        query3 = """select a.name, max(c.relative_velocity_kmph) as maximum_velocity from asteroids a join close_approach c on a.id = c.neo_reference_id group by a.name order by maximum_velocity desc limit 10"""
        df3 = pd.read_sql(query3, suhail)
        st.dataframe(df3)
    elif option == "Potentially hazardous asteroids that have approached Earth more than 3 times":
        query4 = """select a.name, count(c.neo_reference_id) as Approached_Earth_more_than_3_times from asteroids a join close_approach c on a.id = c.neo_reference_id where a.is_potentially_hazardous_asteroid = 1 group by a.name having count(c.neo_reference_id) > 3"""
        df4 = pd.read_sql(query4, suhail)
        st.dataframe(df4)
    elif option == "Month with the most asteroid approaches":
        query5 = """select monthname(close_approach_date) as Month_with_the_most_asteroid_approaches, count(*) as approach_count from close_approach group by Month_with_the_most_asteroid_approaches"""
        df5 = pd.read_sql(query5, suhail)
        st.dataframe(df5)
    elif option == "Asteroid with the fastest ever approach speed":
        query6 = """select a.name, max(c.relative_velocity_kmph) as maximum_speed from asteroids a join close_approach c on a.id = c.neo_reference_id group by a.name order by maximum_speed desc limit 1;"""
        df6 = pd.read_sql(query6, suhail)
        st.dataframe(df6)
    elif option == "Sorted Asteroids in descending order by maximum estimated diameter":
        query7 = """select name, estimated_diameter_max_km from asteroids order by estimated_diameter_max_km desc"""
        df7 = pd.read_sql(query7, suhail)
    elif option == "Asteroid whose closest approach is getting nearer over time":
        query8 = """select a.name, c.miss_distance_km from asteroids a join close_approach c on a.id = c.neo_reference_id order by close_approach_date"""
        df8 = pd.read_sql(query8, suhail)
        st.dataframe(df8)
    elif option == "Asteroids of its closest approach to Earth along with the date and miss distance":
        query9 = """select a.name, c.close_approach_date, c.miss_distance_km from asteroids a join close_approach c on a.id = c.neo_reference_id"""
        df9 = pd.read_sql(query9, suhail)
        st.dataframe(df9)
    elif option == "Asteroids that approached Earth with velocity > 50,000 km/h":
        query10 = """select distinct a.name from asteroids a join close_approach c on a.id = c.neo_reference_id where c.relative_velocity_kmph > 50000"""
        df10 = pd.read_sql(query10, suhail)
        st.dataframe(df10)
    elif option == "Count of asteroids approaches happened per month":
        query11 = """select month(close_approach_date), count(*) as approach_count from close_approach group by month(close_approach_date)"""
        df11 = pd.read_sql(query11, suhail)
        st.dataframe(df11)
    elif option == "Asteroid with the highest brightness":
        query12 = """select name, absolute_magnitude_h from asteroids order by absolute_magnitude_h limit 1"""
        df12 = pd.read_sql(query12, suhail)
        st.dataframe(df12)
    elif option == "Number of hazardous vs non-hazardous asteroids":
        query13 = """select is_potentially_hazardous_asteroid as hazardous_status, count(*) as asteroid_count from asteroids group by is_potentially_hazardous_asteroid"""
        df13 = pd.read_sql(query13, suhail)
        st.dataframe(df13)
    elif option == "Asteroids that passed closer than the Moon (lesser than 1 LD), along with their close approach date and distance":
        query14 = """select a.name, c.close_approach_date, c.miss_distance_lunar from asteroids a join close_approach c on a.id = c.neo_reference_id where c.miss_distance_lunar < 1 order by c.miss_distance_lunar"""
        df14 = pd.read_sql(query14, suhail)
        st.dataframe(df14)
    elif option == "Asteroids that came within 0.05 AU":
        query15 = """select a.name, c.astronomical from asteroids a join close_approach c on a.id = c.neo_reference_id where c.astronomical < 0.05"""
        df15 = pd.read_sql(query15, suhail)
        st.dataframe(df15)

Writing app.py


In [18]:
!wget -q -O - ipv4.icanhazip.com

34.106.210.114


In [19]:
! streamlit run app.py & npx localtunnel --port 8501

⠙

⠹⠸⠼⠴⠦⠧⠇⠏
  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://172.28.0.12:8501
  External URL: http://34.106.210.114:8501

⠋⠙⠹⠸⠼Need to install the following packages:
localtunnel@2.0.2
Ok to proceed? (y) y

⠙⠹⠸⠼⠴⠦⠧⠇⠏⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏⠋⠙⠹your url is: https://loose-schools-refuse.loca.lt
/content/app.py:75: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df1 = pd.read_sql(query1, suhail)
/content/app.py:122: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df13 = pd.read_sql(query13, suhail)
  Stopping...
^C
